[v1.12] Expose memory mapping; Make memfile dump optional

ValentaTomas opened the pull request:
To gain more control over memfile handling during snapshotting, we modify the snapshot logic to optionally skip dumping the memfile.

Also, it adds info about the memory regions when requesting instance info, so we can copy the memory directly.

This is fully compatible with FC v1.12.

This PR is for viewing code changes only and won't be merged.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduce GET /memory and /memory/mappings APIs with VMM support for resident/empty bitmaps and region mappings, and make snapshot mem_file_path optional; update swagger, seccomp, tests, and build/upload tooling.
> 
> - **API/Swagger**
>   - Add `GET /memory` and `GET /memory/mappings` endpoints; wire parsing/response and document in `swagger/firecracker.yaml` (`MemoryResponse`, `MemoryMappingsResponse`, `GuestMemoryRegionMapping`).
>   - Extend `VmmAction` with `GetMemory`/`GetMemoryMappings` and `VmmData` with `Memory`/`MemoryMappings`; pre-boot returns not-supported.
> - **VMM/Core**
>   - Implement memory introspection: `GuestMemoryRegionMapping`, `Vm::guest_memory_mappings`, and `Vm::get_memory_info` (uses `mincore` and zero-page checks) with unified page-size validation.
>   - Extend `InstanceInfo` with optional `memory_regions` field (initialized as `None`).
> - **Snapshotting**
>   - Make `CreateSnapshotParams.mem_file_path` optional and only dump memory when provided; adjust API parsing, runtime handling, and tests.
> - **Seccomp**
>   - Permit `mincore` syscall in `resources/seccomp/*-unknown-linux-musl.json`.
> - **Tests/Tooling**
>   - Add integration tests for new endpoints and a benchmark; update HTTP API client resources.
>   - Add build/upload scripts, Makefile targets, `.tool-versions`, and minor CI script tweak; ignore `.env`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 1dfe77c14f3d5502a9c4e6473a95303c584fabd9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Manciukic committed:
fix(test_reboot): do not assert thread count

An upstream patch was backported to ubuntu 24.04 6.8.0-58 kernel that
makes the nx hugepages recover thread a child of the firecracker
process, thus increasing process count to 7.
As we're not really interested in knowing how many threads we have in
this test, let's remove the assertion altogether.

Signed-off-by: Riccardo Mancini <mancio@amazon.com>

ValentaTomas committed:
Expose memory mapping; Add optional memfile dump

ValentaTomas committed:
Fix compile errors

ValentaTomas commented:
We might also make the snapfile optional/return it as content instead.

ValentaTomas committed:
Remove required field from spec

ValentaTomas committed:
Fix parameter

ValentaTomas committed:
Fix type

ValentaTomas committed:
Add upload script

ValentaTomas committed:
Parse without python

ValentaTomas committed:
Refactor

ValentaTomas committed:
Add tests

ValentaTomas committed:
Fix version

tomassrnka commented:
AI helped review:

- **/memory full scan faults pages in** — `Vm::get_memory_info` (`src/vmm/src/vstate/vm.rs`) walks every page and `memcmp`s each “resident” page; if `mincore` fails it assumes all pages are resident and `memcmp`s them anyway. This can fault nonresident/ballooned pages back in, pull gigabytes into memory, and block the VMM lock for the duration. At minimum, skip `memcmp` unless `mincore` says resident, propagate `mincore` errors, and consider chunking/limits or scanning outside the VMM lock.

- **Long-held VMM lock during scan** — `RuntimeApiController::GetMemory` holds the global VMM mutex while calling `get_memory_info`, so the entire page scan runs under the lock. On large guests this can block other API operations and VM work. Collect mappings under the lock and perform the scan without it, or otherwise limit scan time.

tomassrnka commented:
discussed in IRL, looks good to me

ValentaTomas committed:
Limit test suite

ValentaTomas committed:
Merge branch 'firecracker-v1.12' into firecracker-v1.12-direct-mem

ValentaTomas committed:
Tweak GHA

ValentaTomas committed:
Merge branch 'firecracker-v1.12-direct-mem' of https://github.com/e2b-dev/firecracker into firecracker-v1.12-direct-mem

ValentaTomas committed:
Change runner

ValentaTomas committed:
Remove flaky test invocation

ValentaTomas committed:
Fix build script

ValentaTomas committed:
Switch to 7 hash chars everywhere

ValentaTomas committed:
Remove g

jakubno commented:
bugbot run

jakubno commented on Makefile:
```diff
@@ -0,0 +1,12 @@
+	./scripts/upload.sh $(GCP_PROJECT_ID)
+
+.PHONY: build-and-upload
+make build-and-upload: build upload
```
```suggestion
build-and-upload: build upload
```

jakubno commented on scripts/build.sh:
```diff
@@ -0,0 +1,17 @@
+
+set -euo pipefail
+
+# The format will be: v<major>.<minor>.<patch>_g<commit_hash> — e.g. v1.7.2_g8bb88311
```
```suggestion
# The format will be: v<major>.<minor>.<patch>_g<commit_hash> — e.g. v1.7.2_8bb88311
```

ValentaTomas committed:
Cleanup

bchalios committed:
ci: remove dependency changes test

This is an optional test on the Firecracker side and most of the times
it's ignored (when valid dependency changes happen). Having this fail
blocks our fc-versions releases.

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>

bchalios committed:
api: implement API for dirty memory

Implement API /memory/dirty which returns a bitmap tracking dirty guest
memory. The bitmap is structured as a vector of u64, so its length is:
total_number_of_pages.div_ceil(64).

Pages are ordered in the order of pages as reported by /memory/mappings.

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>

bchalios committed:
feat: enable write-protection on guest memory

UFFD provides an API to enable write-protection for memory ranges
tracked by a userfault file descriptor. Detailed information can be
found here: https://docs.kernel.org/admin-guide/mm/userfaultfd.html.

To use the feature, users need to register the memory region with
UFFDIO_REGISTER_MODE_WP. Then, users need to enable explicitly
write-protection for sub-ranges of the registered region.

Writes in pages within write-protected memory ranges can be handled in
one of two ways. In synchronous mode, writes in a protected page will
cause kernel to send a write protection event over the userfaultfd.
In asynchronous mode, the kernel will automatically handle writes to
protected pages by clearing the write-protection bit. Userspace can
later observe the write protection bit by looking into the corresponding
entry of /proc/<pid>/pagemap.

This commit, uncoditionally, enables write protection for guest memory
using the asynchronous mode.

!NOTE!: asynchronous write protection requires (host) kernel version 6.7
or later).

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>

bchalios committed:
doc(swagger): add /memory/dirty endpoint definition

Add the swagger API definition for the GET /memory/dirty endpoint and
its MemoryDirty response type.

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>

bchalios committed:
doc(swagger): make mem_file_path optional in SnapshotCreateParams

The mem_file_path field is optional when creating a snapshot. If not
specified, the guest memory is not saved to a file.

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>

bchalios committed:
fix(kvmclock): don't pass KVM_CLOCK_REALTIME to KVM_SET_CLOCK

When we pass KVM_CLOCK_REALTIME to KVM_SET_CLOCK, the kernel steps the
monotonic clock by the elapsed wallclock time. This is ok for live
migration, but in the case of snapshots that were take long ago it might
cause issues to guest userspace.

Change the snapshot restore logic to not clear the KVM_CLOCK_REALTIME
bit when restoring kvmclock. Wallclock time will then be handled by
userspace (chronyd or otherwise).

Signed-off-by: Babis Chalios <babis.chalios@e2b.dev>
